File: 0090_subsetsWithDup.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        output = set()
        path = []
        nums.sort()
        self.dfs(nums, output, path)
        return output
        
    def dfs(self, nums, output, path):
        output.add(tuple(path))
        for i in range(len(nums)):
            self.dfs(nums[i+1:], output, path + [nums[i]])

# Alternative solution

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        self.result = []
        self.backtrack(nums, [], 0)
        return self.result

    def backtrack(self, nums, currState, start):
        if currState not in self.result:
            self.result.append(currState)
        for i in range(start, len(nums)):
            self.backtrack(nums, currState + [nums[i]], i + 1)

# Alternative solution

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        self.result = []
        self.backtrack(nums, [], 0)
        return self.result

    def backtrack(self, nums, currState, start):
        self.result.append(currState)
        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i - 1]:
                continue
            self.backtrack(nums, currState + [nums[i]], i + 1)

File: 0001-0500/0090_subsetsWithDup.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> output = new ArrayList<>();
        backtrack(nums, output, new ArrayList<>(), 0);
        return output;
    }

    public void backtrack(int[] nums, List<List<Integer>> output, List<Integer> path, int index) {
        output.add(new ArrayList<>(path));

        if (index == nums.length) {
            return;
        }

        for (int i = index; i < nums.length; i++) {
            if (i > index && nums[i] == nums[i - 1]) {
                continue;
            }
            path.add(nums[i]);
            backtrack(nums, output, path, i + 1);
            path.remove(path.size() - 1);
        }
    }
}

// Alternative solution

class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> output = new ArrayList<>();
        output.add(new ArrayList<>());
        int n = nums.length;
        int end = 0;

        for (int i = 0; i < n; i++) {
            int start = 0;
            if (i > 0 && nums[i] == nums[i - 1]) {
                start = end + 1;
            }
            int k = output.size();
            end = k - 1;

            for (int j = start; j < end + 1; j++) {
                List<Integer> list = new ArrayList<>(output.get(j));
                list.add(nums[i]);
                output.add(list);
            }
        }

        return output;
    }
}
